staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -476,6 +476,11 @@ func (h *heartbeat) start() {
 				klog.V(8).Infof("Websocket Ping succeeeded")
 			} else {
 				klog.Errorf("Websocket Ping failed: %v", err)
+				if gwebsocket.IsUnexpectedCloseError(err, gwebsocket.CloseGoingAway, gwebsocket.CloseAbnormalClosure) {
```

Comment:
The key here is to identify the errors that are not retriable, all Close errors are not retriable, right? if the connection is closed we should exit, independently of the h.closer to be closed or not.

I think we should implement this logic here

https://github.com/gorilla/websocket/blob/9111bb834a68b893cebbbaed5060bdbc1d9ab7d2/conn.go#L1161-L1166
```go
			err := c.WriteControl(PongMessage, []byte(message), time.Now().Add(writeWait))
			if err == ErrCloseSent {
				return nil
			} else if e, ok := err.(net.Error); ok && e.Temporary() {
				return nil
			}
			return err
```
and only retry if is temporary or ErrCloseSent

Reply:
This `e.Temporary()` looks like a useful function

Reply:
is not redundant, is a race, and the code must handle both

Reply:
hmm.. `ErrCloseSent` does not cover what we are looking for https://github.com/gorilla/websocket/blob/666c197fc9157896b57515c3a3326c3f8c8319fe/conn.go#L87 (because abnormal closure's message is different https://github.com/gorilla/websocket/blob/666c197fc9157896b57515c3a3326c3f8c8319fe/conn.go#L113) and `temporary` function seems deprecated https://github.com/golang/go/blob/774d001d178569283cb719bd062cf5d18de51f47/src/net/net.go#L397-L400

Reply:
I just hinted the same logic gorilla is using under the hood for the same problem, otherwise we just return in any error ... swallowing the errors like now is wrong as we can see

Reply:
the goal here is to make a "isRetriableError(err error) bool" function and exit if is not retriable, I think taht the existing gorilla code to handle the "WriteControl" message errors is a good starting point